### PR TITLE
Add actions for debugger rules and set up PR CI for unit tests

### PR DESCRIPTION
**Pull request by ndodda-amazon**
*Issue #, if available:*

*Description of changes:*
Add action classes to be used by the sagemaker SDK to simplify the usage of actions to be invoked when debugger rules fire during training.

Tested these changes by:
- Building a rules config wheel with these changes
- Building a sagemaker SDK wheel with the changes in [this branch](https://github.com/ndodda-amazon/sagemaker-python-sdk/tree/actions).
- Using a training script that sleeps for 20 minutes after training (before the script ends)
- Running a training job with stalled training rule with threshold 15 minutes and all three of the new actions.
- Confirmed that the stalled training rule fired, the training job was stopped early, and both an email/SMS was sent to the email address/phone number specified in the actions.

See code block below for user experience:

```
actions = rule_configs.ActionList(rule_configs.StopTraining(), rule_configs.Email("abc@abc.com"), rule_configs.SMS("+1234567890"))
rules = [
    Rule.sagemaker(rule_configs.vanishing_gradient()), 
    Rule.sagemaker(rule_configs.loss_not_decreasing()),
    Rule.sagemaker(
        rule_configs.stalled_training_rule(), 
        rule_parameters={"threshold": str(60*15)},
        actions=actions
    )
]


estimator = TensorFlow(
    role=sagemaker.get_execution_role(),
    base_job_name='test-actions',
    instance_count=1,
    instance_type='ml.c5.4xlarge',
    framework_version='1.15',
    py_version='py3',
    source_dir="src",
    entry_point="simple_stalled_training.py",
    rules = rules
)
```

Also added unit tests for these changes and set up the infrastructure to run these unit tests as a part of the PR CI. Since the codebase itself is framework independent, we can have a separate build for just running unit tests (and any future builds would only run sagemaker tests for the desired framework + version).

TODO:
- Add end to end sagemaker tests for these changes
- Set up code coverage

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by NihalHarish** [comment]
What is the delta to set up a CI for this?

**Comment by ndodda-amazon** [comment]
> What is the delta to set up a CI for this?

Days. Let's assume we just want to test the changes in this PR. This would require:
- unit tests to test various combinations of input for actions, including bad input to verify that the validation is correctly catching issues
- sagemaker tests to test various combinations of actions with various of stalled training rule thresholds (for testing cases where the rule fires or not). to be comprehensive, these sagemaker tests should be created for all frameworks + versions we support.
- creating builds and buildspecs for each framework + version to test. for the CI, we can simply run the tests for latest versions of each framework that we support.

Major problem is that these changes are essentially built on top off of debugger rules by definition. So just simply adding tests for actions isn't that useful when we don't even have tests + CI for debugger rules to start with. If action tests fail, is it an issue with stalled training rule or the action itself?

**Review by Vikas-kum — COMMENTED** [review]
Can you write the estimator code that you had to write to include actions in detail section.

**Comment by ndodda-amazon** [comment]
> Can you write the estimator code that you had to write to include actions in detail section.

Added to the description

**Review by Vikas-kum — APPROVED** [review]
Looks good to me. Verify the experience with Satadal once.

**Comment by ndodda-amazon** [comment]
Added unit tests and set up PR CI to run these unit tests. This required minor changes to rules config actions, so I reran the manual sagemaker test by rebuilding the binaries and confirmed that the actions were still invoked correctly.

File: tests/core/actions/utils.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    Note that we need the exec call to mimic the same behavior in the rules container.
+
+    If this triggers a syntax error, the exec call is set up incorrectly and needs to be fixed.
+    If this triggers a JSON decode error, the action string is badly formatted.
+    If this triggers an assertion error, the deserialized action JSON does not match the original action parameters.
+    """
+    try:
+        exec(f'import json; assert json.loads("{action_str}") == {action_parameters}')
+    except (SyntaxError, json.JSONDecodeError, AssertionError) as e:
+        raise Exception(f"Error occurred during action string validation.")
```
Comment (Vikas-kum) [inline tests/core/actions/utils.py:16]: more details about exception, you should bubble up e.

File: smdebug_rulesconfig/actions/actions.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+        :return: Action parameters serialized as a string dictionary.
+        """
+        return (
+            "{"
+            + ", ".join(
+                [f'\\"{key}\\": \\"{value}\\"' for key, value in self.action_parameters.items()]
+            )
+            + "}"
+        )
```
Comment (Vikas-kum) [inline smdebug_rulesconfig/actions/actions.py:30]: you should validate this json str before returning using your validate method and raise appropriate exception message if it is invalid.

Same with other serialize methods.

**Review by Vikas-kum — CHANGES_REQUESTED** [review]

File: tests/core/actions/utils.py (hunk 1 above)
Comment (ndodda-amazon) [reply tests/core/actions/utils.py:16]: e doesn't provide a useful message, it just lists the exception (since the code is triggered in another process)

File: smdebug_rulesconfig/actions/actions.py (hunk 2 above)
Comment (ndodda-amazon) [reply smdebug_rulesconfig/actions/actions.py:30]: The arguments (`self.action_parameters`) are validated by this point, and the json str is formed by us. We have tests that validate this json str. User should not be facing exception with json str.

File: smdebug_rulesconfig/actions/actions.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+            to one active training job, otherwise no training job will be stopped.
+        """
+        if training_job_prefix is not None:
+            validate_string(
+                "training_job_prefix", training_job_prefix
+            ), "training_job_prefix must be a string!"
+        self.use_default_training_job_prefix = True if training_job_prefix is None else False
+        super().__init__(training_job_prefix=training_job_prefix)
+
+    def update_training_job_prefix_if_not_specified(self, training_job_name):
```
[the remark this answers is not in the document]
Comment (ndodda-amazon) [reply smdebug_rulesconfig/actions/actions.py:95]: sagemaker SDK, see the docstring.

File: tests/core/actions/utils.py (hunk 1 above)
Comment (ndodda-amazon) [reply tests/core/actions/utils.py:16]: hence why I updated the docstring about what each possible exception means.

File: smdebug_rulesconfig/actions/actions.py [hunk 4]
```diff
@@ -39,7 +42,7 @@ def __init__(self, *actions):
             isinstance(action, Action) for action in actions
         ), "actions must be list of Action objects!"
 
-        self.action_parameters = actions
+        self.actions = actions
 
     def update_training_job_prefix_if_not_specified(self, training_job_name):
```
[the remark this answers is not in the document]
Comment (ndodda-amazon) [reply smdebug_rulesconfig/actions/actions.py:47]: sagemaker SDK, see the docstring.

File: config/buildspec_core.yml [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+phases:
+  install:
+    commands:
+        -  . config/change_branch.sh
+        - su && apt-get update
+        - apt-get install sudo -qq -o=Dpkg::Use-Pty=0 # silence output: https://askubuntu.com/a/668859/724247
+        - sudo apt-get update -qq -o=Dpkg::Use-Pty=0
+        - sudo apt-get install unzip -qq -o=Dpkg::Use-Pty=0
+        - pip install --upgrade pip==19.3.1
+        - pip install -q pytest wheel pytest-html pre-commit awscli pytest-cov
```
[the remark this answers is not in the document]
Comment (ndodda-amazon) [reply config/buildspec_core.yml:19]: copied this from: https://github.com/awslabs/sagemaker-debugger/blob/master/config/buildspec.yml#L35

yes I can pin versions here, but let's make sure we do the same for sagemaker-debugger too.

File: smdebug_rulesconfig/actions/actions.py (hunk 2 above)
Comment (Vikas-kum) [reply smdebug_rulesconfig/actions/actions.py:30]: playing devils advocate here- What if i provide a string like vikas\"\@gmail.com (some escaped string which makes your json creation invalid)? 
validating string means the json string that we created should be valid, otherwise it is going to fail later in rules, which we want to avoid as early as possible. We can avoid that by makign sure that the json that we created out of parameters are indeed valid, if not we raise appropriate errors in SDK itself.

File: tests/core/actions/utils.py (hunk 1 above)
Comment (Vikas-kum) [reply tests/core/actions/utils.py:16]: do we need to do 'exec'? 
Its all pythonic statement, it can be verified in same process, No?

File: config/buildspec_core.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (Vikas-kum) [reply config/buildspec_core.yml]: That's right. We should always pin the version in all our builds.

File: tests/core/actions/utils.py (hunk 1 above)
Comment (ndodda-amazon) [reply tests/core/actions/utils.py:16]: See the docstring. The point is to mimic the same behavior in the rules container, where the JSON is parsed within an exec call.

The reason why this matters is because two backslashes are needed for strings if this is executed within an exec call. By simply parsing json in this process, this triggers json validation error.

File: smdebug_rulesconfig/actions/actions.py (hunk 2 above)
Comment (ndodda-amazon) [reply smdebug_rulesconfig/actions/actions.py:30]: Makes sense. I think this falls under the argument validation though, so I'll update that to be more strict.

File: tests/core/actions/utils.py (hunk 1 above)
Comment (Vikas-kum) [reply tests/core/actions/utils.py:16]: Makes sense.

File: smdebug_rulesconfig/actions/actions.py (hunk 2 above)
Comment (Vikas-kum) [reply smdebug_rulesconfig/actions/actions.py:30]: Should you call the same function, that you wrote in test/core/utils in this PR? 
 Ultimately you want is the end string that serialize is returning, is valid as seen by rules container. 
 Also key thing would be how do we raise proper error mesg for customer, so that mesg clearly specifies how can it be corrected.

Comment (ndodda-amazon) [reply smdebug_rulesconfig/actions/actions.py:30]: We should have strict validation on any user arguments so that ideally we don't need to rely on that function to catch errors from the user. Regarding that function I wrote, we would need to raise a proper error message for customer as you mentioned. And we can't, as I mentioned in a separate comment thread.

So the best solution is to use strict argument validation (regex) to catch all potential errors. Perhaps we should still call that function anyway since user experience is better if error is caught early on even if there is no helpful message.

File: config/buildspec_core.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (ndodda-amazon) [reply config/buildspec_core.yml]: Done.

File: smdebug_rulesconfig/actions/actions.py (hunk 2 above)
Comment (Vikas-kum) [reply smdebug_rulesconfig/actions/actions.py:30]: >So the best solution is to use strict argument validation (regex) to catch all potential errors. Perhaps we should still call that function anyway since user experience is better if error is caught early on even if there is no helpful message.

Yeah i agree. 
so, 
validate_params()
create_json
validate json

File: smdebug_rulesconfig/actions/actions.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+        }
+        ```
+
+        :param phone_number: Phone number to send the SMS to.
+        """
+        validate_string("phone_number", phone_number)
+        super().__init__(endpoint=phone_number)
+
+
+def is_valid_action_object(actions):
```
Comment (metrizable) [inline smdebug_rulesconfig/actions/actions.py:169]: is this method used anywhere but the tests?

**Review by metrizable — CHANGES_REQUESTED** [review]

File: smdebug_rulesconfig/actions/actions.py (hunk 2 above)
Comment (ndodda-amazon) [reply smdebug_rulesconfig/actions/actions.py:38]: Done.

File: tests/core/actions/utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ndodda-amazon) [reply tests/core/actions/utils.py]: Already discussed here: https://github.com/awslabs/sagemaker-debugger-rulesconfig/pull/32#discussion_r545401362

TL;DR can't get the traceback since the call was made in exec. All we can get is the type of exception (unless we put in more work to capture output). Docstring has complete information about why validation could fail.

[the remark this answers is not in the document]
Comment (metrizable) [reply tests/core/actions/utils.py]: any reason you wouldn't `raise..from`?

or at the very least include the exception `e` in the newly raised exception? as it stands, the `f-string` isn't used, and `e` isn't used.

File: smdebug_rulesconfig/actions/actions.py (hunk 6 above)
Comment (ndodda-amazon) [reply smdebug_rulesconfig/actions/actions.py:176]: In sagemaker SDK, see the docstring.

File: .pre-commit-config.yaml [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+    -   id: trailing-whitespace
+
+- repo: https://github.com/humitos/mirrors-autoflake.git
+  rev: v1.3
+  hooks:
+    - id: autoflake
+      args: ['--in-place', '--expand-star-imports', '--ignore-init-module-imports', '--remove-all-unused-imports']
+
+-   repo: https://github.com/psf/black
+    rev: 19.3b0
```
Comment (metrizable) [inline .pre-commit-config.yaml:16]: do you want to use up-to-date `20.8b1`? `19.3b0` is from March 2019

File: smdebug_rulesconfig/__init__.py [hunk 8]
```diff
@@ -1,3 +1,5 @@
+from .actions.actions import SMS, ActionList, Email, StopTraining, is_valid_action_object
```
[the remark this answers is not in the document]
Comment (ndodda-amazon) [reply smdebug_rulesconfig/__init__.py:1]: We use pre-commit to handle style formatting automatically when committing code. This is consistent with what we have in our other repositories.

File: config/change_branch.sh [hunk 9]
```diff
@@ -0,0 +1,42 @@
+export CODEBUILD_GIT_BRANCH="$(git symbolic-ref HEAD --short 2>/dev/null)"
+if [ "$CODEBUILD_GIT_BRANCH" = "" ] ; then
+  CODEBUILD_GIT_BRANCH="$(git branch -a --contains HEAD | sed -n 2p | awk '{ printf $1 }')";
+  export CODEBUILD_GIT_BRANCH=${CODEBUILD_GIT_BRANCH#remotes/origin/};
+fi
+
+cd $CODEBUILD_SRC_DIR && git checkout $CODEBUILD_GIT_BRANCH
```
Comment (metrizable) [inline config/change_branch.sh:7]:
```suggestion
pushd $CODEBUILD_SRC_DIR && git checkout $CODEBUILD_GIT_BRANCH
```

File: config/change_branch.sh [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+export CURRENT_COMMIT_HASH=$(git log -1 --pretty=%h);
+#export CURRENT_COMMIT_DATE="$(git show -s --format=%ci | cut -d' ' -f 1)$(git show -s --format=%ci | cut -d' ' -f 2)";
+export CURRENT_DATETIME=$(date +'%Y%m%d_%H%M%S')
+export CURRENT_REPO_NAME=$(basename `git rev-parse --show-toplevel`) ;
+export CURRENT_COMMIT_PATH="$CURRENT_DATETIME/$CURRENT_COMMIT_HASH"
+cd ..
```
Comment (metrizable) [inline config/change_branch.sh:13]:
```suggestion
popd
```
safer to do `pushd..popd` if some commands in between change the working directory

File: config/tests.sh [hunk 11]
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env bash
+# TODO: Add end to end sagemaker tests for each framework and version so that they can be run here.
+
+set -e
+set -ex
+set -o pipefail
```
[the remark this answers is not in the document]
Comment (ndodda-amazon) [reply config/tests.sh]: Thanks. Worked off of existing `tests.sh` from [sagemaker-debugger](https://github.com/awslabs/sagemaker-debugger/blob/master/config/tests.sh#L3) so we should update that as well.

File: config/tests.sh [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+
+
+run_for_framework() {
+    python -m pytest ${code_coverage_smdebug:+--cov=./ --cov-append}  --durations=50 --html=$REPORT_DIR/report_$1.html -v -s --self-contained-html tests/$1
+}
```
Comment (metrizable) [inline config/tests.sh:9]: have you considered using `tox`?

**Review by metrizable — CHANGES_REQUESTED** [review]

File: config/change_branch.sh [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+
+export CODEBUILD_ACCOUNT_ID=$(aws sts get-caller-identity --query 'Account' --output text)
+export CODEBUILD_PROJECT=${CODEBUILD_BUILD_ID%:$CODEBUILD_LOG_PATH}
+
+export CODEBUILD_BUILD_URL=https://$AWS_DEFAULT_REGION.console.aws.amazon.com/codebuild/home?region=$AWS_DEFAULT_REGION#/builds/$CODEBUILD_BUILD_ID/view/new
+
+echo "INFO =============================BUILD STARTED==================================="
+echo "INFO =============================Build details========================== ::"
+echo "INFO CODEBUILD_CURRENT_BUILD_URL = $CODEBUILD_BUILD_URL"
+#echo "INFO CURRENT_REPO_NAME = $CURRENT_REPO_NAME"
```
Comment (NihalHarish) [inline config/change_branch.sh:37]: nit: is there a reason you've kept these comments?

File: tests/core/actions/utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ndodda-amazon) [reply tests/core/actions/utils.py]: We are using the `e`, see [here](https://github.com/awslabs/sagemaker-debugger-rulesconfig/pull/32/files#diff-056dd7e56fcbd0d7105a6b9c7c2b8b321e8f191254456c20168b34d0cf1783e3R51).

I think you might've reviewed before I pushed this change.

File: smdebug_rulesconfig/actions/actions.py [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+            self.serialize(), self.action_parameters
+        )  # sanity check, not expected to error!
+
+    def serialize(self):
+        """
+        Serialize the action parameters as a string dictionary.
+
+        :return: Action parameters serialized as a string dictionary.
+        """
+        return (
```
Comment (NihalHarish) [inline smdebug_rulesconfig/actions/actions.py:32]: Can we use the standard json library to handle serialization?

File: config/buildspec_core.yml [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+        - pip install -q pytest==6.1.2 pytest-cov==2.10.1 wheel==0.35.1 pytest-html==3.0.0 pre-commit==2.6.0 awscli==1.18.199
+
+  pre_build:
+    commands:
+      - cd $CODEBUILD_SRC_DIR && pre-commit install && pre-commit run --all-files
+
+  build:
+    commands:
+      - cd $CODEBUILD_SRC_DIR  && python setup.py bdist_wheel --universal
+      - cd $CODEBUILD_SRC_DIR  && chmod +x config/tests.sh && PYTHONPATH=. && ./config/tests.sh  && mkdir -p upload/$CURRENT_COMMIT_PATH/wheels && cp ./dist/*.whl upload/$CURRENT_COMMIT_PATH/wheels && cd ..
```
Comment (metrizable) [inline config/buildspec_core.yml:28]: will you also release a source distribution?

**Review by metrizable — CHANGES_REQUESTED** [review]

File: smdebug_rulesconfig/actions/actions.py [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+        action parameters and drops keys whose values are `None`.
+
+        :param action_parameters: Dictionary of action parameters.
+        """
+        action_parameters["name"] = self.__class__.__name__.lower()
+        self.action_parameters = {
+            key: value for key, value in action_parameters.items() if value is not None
+        }
+
+    def serialize(self):
```
[the remark this answers is not in the document]
Comment (ndodda-amazon) [reply smdebug_rulesconfig/actions/actions.py:26]: See [this thread](https://github.com/awslabs/sagemaker-debugger-rulesconfig/pull/32#discussion_r545402799).

The following serialization is not technically valid JSON if parsed as is. It needs to be structured in a way where it will be parsed correctly (`json.loads` will eventually be called on the string within an `exec` call)

File: smdebug_rulesconfig/actions/actions.py [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+            "{"
+            + ", ".join(
+                [f'\\"{key}\\": \\"{value}\\"' for key, value in self.action_parameters.items()]
+            )
+            + "}"
+        )
+
+
+class ActionList(object):
+    def __init__(self, *actions):
```
Comment (NihalHarish) [inline smdebug_rulesconfig/actions/actions.py:42]: Can you add a type annotation for this parameter.

`: List[Action]`

File: smdebug_rulesconfig/actions/actions.py [hunk 18]
```diff
@@ -0,0 +1,184 @@
+from smdebug_rulesconfig.actions.utils import (
+    validate_training_job_prefix,
+    validate_email_address,
+    validate_phone_number,
+    validate_action_str,
+)
+
+
+class Action(object):
+    def __init__(self, **action_parameters):
```
Comment (NihalHarish) [inline smdebug_rulesconfig/actions/actions.py:10]: nit: Add a type annotation here

File: .pre-commit-config.yaml (hunk 7 above)
Comment (ndodda-amazon) [reply .pre-commit-config.yaml:16]: Appreciate the comments on the pre-commit/tests.sh/buildspec, but these are simply mimicking what is done in our other repos. Making all these changes here would result in inconsistent behavior for the other repos. I will keep track of all these suggestions so that when the time comes for us to update all these files, we will do this for all of our repos at the same time.

Comment (ndodda-amazon) [reply .pre-commit-config.yaml:16]: Going to resolve this and related comments.

File: smdebug_rulesconfig/actions/actions.py [hunk 19]
```diff
@@ -0,0 +1,184 @@
...
+        ```
+
+        :param email_address: Email address to send the email notification to.
+        """
+        validate_email_address("email_address", email_address)
+        super().__init__(endpoint=email_address)
+
+
+class SMS(Action):
+    def __init__(self, phone_number):
```
Comment (NihalHarish) [inline smdebug_rulesconfig/actions/actions.py:147]: Type annotate here. It isn't clear if the param is a str or int

File: config/change_branch.sh (hunk 9 above)
Comment (ndodda-amazon) [reply config/change_branch.sh:7]: Same as above.

File: config/change_branch.sh (hunk 10 above)
Comment (ndodda-amazon) [reply config/change_branch.sh:13]: Same as above.

File: config/tests.sh (hunk 12 above)
Comment (ndodda-amazon) [reply config/tests.sh:9]: Same as above

File: smdebug_rulesconfig/actions/actions.py [hunk 20]
```diff
@@ -0,0 +1,184 @@
...
+                        "sns:CreateTopic",
+                        "sns:Subscribe"
+                    ],
+                    "Resource": "arn:aws:sns:*:<account-id>:SMDebugRules"
+                }
+            ]
+        }
+        ```
+
+        :param phone_number: Phone number to send the SMS to.
```
Comment (NihalHarish) [inline smdebug_rulesconfig/actions/actions.py:170]: indicate type in this docstring since we expect customers to run 
`help` for this object.

File: config/buildspec_core.yml (hunk 15 above)
Comment (ndodda-amazon) [reply config/buildspec_core.yml:28]: Your question seems unrelated to this change. But yes, when we release to PyPi, we will also release a source distribution.

**Review by NihalHarish — CHANGES_REQUESTED** [review]

File: config/change_branch.sh (hunk 13 above)
Comment (ndodda-amazon) [reply config/change_branch.sh:37]: Copied from sagemaker-debugger.

File: smdebug_rulesconfig/actions/actions.py (hunk 14 above)
Comment (ndodda-amazon) [reply smdebug_rulesconfig/actions/actions.py:32]: No, see: https://github.com/awslabs/sagemaker-debugger-rulesconfig/pull/32#discussion_r545402799

**Review by Vikas-kum — APPROVED** [review]

File: smdebug_rulesconfig/actions/actions.py (hunk 17 above)
Comment (ndodda-amazon) [reply smdebug_rulesconfig/actions/actions.py:42]: Apparently this should just be the type in the list: https://www.python.org/dev/peps/pep-0484/#arbitrary-argument-lists-and-default-argument-values.

So `: Action`.

Comment (ndodda-amazon) [reply smdebug_rulesconfig/actions/actions.py]: Done.

File: smdebug_rulesconfig/actions/actions.py (hunk 18 above)
Comment (ndodda-amazon) [reply smdebug_rulesconfig/actions/actions.py]: Done.

File: smdebug_rulesconfig/actions/actions.py (hunk 19 above)
Comment (ndodda-amazon) [reply smdebug_rulesconfig/actions/actions.py]: Done.

File: smdebug_rulesconfig/actions/actions.py (hunk 20 above)
Comment (ndodda-amazon) [reply smdebug_rulesconfig/actions/actions.py:171]: I'll make this more clear for now but Miyoung will formally be updating dostrings in a followup PR.

File: smdebug_rulesconfig/__init__.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ndodda-amazon) [reply smdebug_rulesconfig/__init__.py:1]: Will keep track of this suggestion, we should update these YAMLs since they're pretty old.

File: smdebug_rulesconfig/actions/actions.py [hunk 21]
```diff
@@ -0,0 +1,177 @@
...
+
+class ActionList(object):
+    def __init__(self, *actions):
+        """
+        Higher level object to maintain a list of actions to be invoked when a rule is fired. Offers higher level
+        `serialize` function to handle serialization of actions as a string list of dictionaries.
+
+        :param actions: List of actions.
+        """
+        assert all(
```
[the remark this answers is not in the document]
Comment (ndodda-amazon) [reply smdebug_rulesconfig/actions/actions.py]: Good point, updated all assertions to to raise actual exceptions with the error message.

Again need to keep track of this since we don't have the same behavior across repos.

File: smdebug_rulesconfig/actions/actions.py [hunk 22]
```diff
@@ -0,0 +1,177 @@
...
+
+        :param training_job_prefix: The prefix of the training job to stop if the rule is fired. This must only refer
+            to one active training job, otherwise no training job will be stopped.
+        """
+        if training_job_prefix is not None:
+            validate_string(
+                "training_job_prefix", training_job_prefix
+            ), "training_job_prefix must be a string!"
+        self.use_default_training_job_prefix = True if training_job_prefix is None else False
+        super().__init__(training_job_prefix=training_job_prefix)
```
[the remark this answers is not in the document]
Comment (ndodda-amazon) [reply smdebug_rulesconfig/actions/actions.py]: Done. We should make this behavior consistent across all of our repos.

File: smdebug_rulesconfig/actions/actions.py [hunk 23]
```diff
@@ -168,7 +168,7 @@ def __init__(self, phone_number: str):
         }
         ```
 
-        :param phone_number: Phone number to send the SMS to.
+        :param phone_number: 10 digit phone number preceded by a `+`.
```
Comment (Vikas-kum) [inline smdebug_rulesconfig/actions/actions.py:171]: is +10 digit too strict? can it be something else out of United States ?

Comment (ndodda-amazon) [reply smdebug_rulesconfig/actions/actions.py:171]: good catch, I'll update this so that it can accept a number from 10-15 digits (can a valid phone number be longer or shorter?)

Comment (ndodda-amazon) [reply smdebug_rulesconfig/actions/actions.py]: Done.

**Review by NihalHarish — APPROVED** [review]
